tools/jenkins-proxy/proxy.go

```diff
@@ -0,0 +1,303 @@
+	var err error
+	backoff := 100 * time.Millisecond
+	for retries := 0; retries < maxRetries; retries++ {
+		resp, err = c.doRequest(method, path, body, h)
+		if err == nil && resp.StatusCode < 500 {
+			break
+		} else if err == nil {
+			resp.Body.Close()
+		}
+
+		time.Sleep(backoff)
+		backoff *= 2
+	}
+	return resp, err
+}
+
+func (p *Proxy) doRequest(method, path string, body io.Reader, h *http.Header) (*http.Response, error) {
+	req, err := http.NewRequest(method, path, body)
+	if err != nil {
+		return nil, err
+	}
+	if h != nil {
+		copyHeader(h, &req.Header)
+	}
+	// Configure auth
+	for _, m := range p.Masters {
+		if strings.HasPrefix(path, m.url.String()) {
+			if m.Auth.Basic != nil {
+				req.SetBasicAuth(m.Auth.Basic.User, m.Auth.Basic.token)
+			} else if m.Auth.BearerToken != nil {
+				req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", m.Auth.BearerToken.token))
+			}
+			break
+		}
+	}
+	return p.client.Do(req)
+}
+
+// TODO: Prometheus metrics
+func (p *Proxy) handle(w http.ResponseWriter, r *http.Request) {
+	log.Print(r.Method + ": " + r.URL.String())
+	w.Header().Set("X-Jenkins-Proxy", "JenkinsProxy")
+
+	if p.ProxyAuth != nil {
+		user, pass, ok := r.BasicAuth()
+		if !ok {
+			http.Error(w, "Basic authentication required.", http.StatusUnauthorized)
+			return
+		}
+		userCmp := subtle.ConstantTimeCompare([]byte(p.ProxyAuth.User), []byte(user))
+		passCmp := subtle.ConstantTimeCompare([]byte(p.ProxyAuth.token), []byte(pass))
+		if userCmp != 1 || passCmp != 1 {
+			http.Error(w, "Basic authentication failed.", http.StatusUnauthorized)
+			return
+		}
+
+	}
+
+	// There are three different kinds of requests that the jenkins-operator
+	// is doing:
+	//
+	// * build requests (needs a job)
+	// * status requests (needs a job)
+	// * queue requests (does not need a job)
+	//
+	// Requests that need a job will need to be resolved using the proxy
+	// cache. Today, queue requests will always contain the correct hostname
+	// because the operator assigns the queue url it gets back from a build
+	// request on a prow job. Eventually, this will change (see
+	// https://github.com/kubernetes/test-infra/issues/4366) and then we should
+	// broadcast queue requests to all masters.
+	parts := strings.Split(r.URL.Path, "/")
+	jobIndex := -1
+	for i, part := range parts {
+		if part == "job" {
+			// This is a job-specific request. Record the index.
+			jobIndex = i + 1
+			break
+		}
+	}
+	// If this is not a job-specific request, fail for now. Eventually we
+	// are going to proxy queue requests.
+	if jobIndex == -1 {
+		http.Error(w, "Forbidden.", http.StatusForbidden)
+		return
+	}
+	// Sanity check
+	if jobIndex+1 > len(parts) {
+		http.Error(w, "Forbidden.", http.StatusForbidden)
+		return
+	}
+
+	requestedJob := parts[jobIndex]
+	// Get the destination URL by looking at the job cache.
+	destURL := p.getDestURL(requestedJob)
+	if len(destURL) == 0 {
+		// Update the cache by relisting from all masters.
```

Comment:
what are we gaining from this cache if we relist on every access?

Reply:
Newly added jobs in Jenkins can be picked up on the fly w/o the need to restart the proxy.

Reply:
*We are relisting on cache misses

Reply:
Oh, right, sorry misread that

Reply:
Oh, right, sorry misread that